tensorpipe/test/peer_group.h

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#pragma once
+
+#include <string>
+#include <thread>
+
+#include <unistd.h>
+
+#include <gtest/gtest.h>
+
+class PeerGroup {
+ public:
+  static constexpr int kServer = 0;
+  static constexpr int kClient = 1;
+
+  virtual ~PeerGroup() = default;
+  virtual void sendPeer(int peerId, const std::string&) = 0;
+  virtual std::string recvPeer(int peerId) = 0;
+  virtual void spawn(std::function<void()>, std::function<void()>) = 0;
+
+  void donePeer(int peerId) {
+    sendPeer(1 - peerId, kDone);
+    std::unique_lock<std::mutex> lock(m_);
+    done_[peerId] = true;
+    cond_var_[peerId].notify_one();
+  }
+
+  void joinPeer(int peerId) {
+    while (recvPeer(peerId) != kDone)
+      ;
+    std::unique_lock<std::mutex> lock(m_);
+    while (!done_[peerId]) {
+      cond_var_[peerId].wait(lock);
+    }
+  }
+
+ private:
+  const std::string kDone = "done";
+  std::mutex m_;
+  bool done_[2] = {false, false};
+  std::condition_variable cond_var_[2];
+};
+
+class ThreadPeerGroup : public PeerGroup {
+ public:
+  void sendPeer(int peerId, const std::string& str) override {
+    q_[peerId].push(str);
+  }
+
+  std::string recvPeer(int peerId) override {
+    return q_[peerId].pop();
+  }
+
+  void spawn(std::function<void()> f1, std::function<void()> f2) override {
+    std::function<void()> fns[] = {std::move(f1), std::move(f2)};
+    std::vector<std::thread> ts;
+
+    for (auto& f : fns) {
+      ts.emplace_back(f);
+    }
+
+    for (auto& t : ts) {
+      t.join();
+    }
+  }
+
+ private:
+  tensorpipe::Queue<std::string> q_[2];
+};
+
+class ProcessPeerGroup : public PeerGroup {
+ public:
+  ProcessPeerGroup() {
+    for (int i = 0; i < 2; ++i) {
+      pipe(pipefd[i]);
+    }
+  }
+
+  ~ProcessPeerGroup() override {
+    for (int i = 0; i < 2; ++i) {
+      for (int j = 0; j < 2; ++j) {
+        close(pipefd[i][j]);
+      }
+    }
+  }
+
+  void sendPeer(int peerId, const std::string& str) override {
+    write(pipefd[peerId][1], str.c_str(), str.length());
+    write(pipefd[peerId][1], "\n", 1);
+  }
+
+  std::string recvPeer(int peerId) override {
+    std::string str;
+    char c;
+    while (read(pipefd[peerId][0], &c, 1) > 0) {
```

Comment:
It probably makes sense to check the return code of the `read` and `write` calls and if they signal an error print a nice readable message. And, in fact, we should probably also do it for `pipe` (above) and `fork` (below).